README.md
# flask-quickstart
**[v1.1.2](https://github.com/bittobennichan/flask-quickstart/blob/master/CHANGES)**

**flask-quickstart** is a minimal boilerplate/template for starting your new Flask project.

The good thing about *flask-quickstart* is that it does not assume that you are using any Flask extensions. It sets up your project layout, main *Blueprint*, configuration and then gets out of your way!

## Quickstart to flask-quickstart

First clone the repostiory

```
$ git clone https://github.com/bittobennichan/flask-quickstart.git
```
Change directory to flask-quickstart

```
$ cd flask-quickstart
```
Create a virtual environment and install dependencies
```
$ python -m venv venv
$ source venv/bin/activate
$ python -m pip install --upgrade pip setuptools
$ python -m pip install -r requirements.txt
```

Start the development server
```
$ python manage.py runserver
```
That's it! 

Navigate to `http://127.0.0.1:5000/` and start your new project!

## Project layout
```
flask-quickstart/
├── app/
│   ├── __init__.py
│   ├── main/
│   │   ├── forms.py
│   │   ├── __init__.py
│   │   └── views.py
│   ├── models/
│   ├── static/
│   │   ├── css/
│   │   ├── img/
│   │   └── js/
│   └── templates/
│       ├── common/
│       ├── error/
│       │   ├── 404.html
│       │   └── 500.html
│       └── main/
│           └── index.html
├── CHANGES
├── config.py
├── instance/
│   └── config.py
├── LICENSE
├── manage.py
├── README.md
├── requirements.txt
├── tests/
│   ├── __init__.py
│   └── test_main.py
└── wsgi.py
```

When you clone the repo, you'll notice that you will be missing the *config.py* file in instance folder. You will find an *example_config.py* file instead.

```
├── instance/
    └── example_config.py
```

This is b'coz it should be set to be ignored in the *.gitignore* file. Infact everything except *example_config.py* is set to be ignored. This is where you will store the deployment/production configuration values. You don't want to accidently push this to Github. Rename the file to *config.py* after you clone the repo.


If available, the values in `flask-quickstart/instance/config.py` will override any config values set in tthe environment form `flask-quickstart/config.py` when environment is set to `production` (it is set to `development` by default).

The **wsgi.py** file uses `production` as the environment. This file is to be used by WSGI servers such as Gunicorn.

You can also test production environment using `manage.py`

```
$ python manage.py runserver --env production
```

You'll find that there are `.gitkeep` files in some folders. Git does not track empty folders. This is a convention used to preserve the folder structure. These files are not necessary and you can delete them if you want.

## manage.py

As mentioned earlier, **manage.py** has just one command **runserver** which is used to run the Flask development server.

It takes two optional parameters

- `env` - Environment to use while running server. Supports  `production`, `development` and `testing` out of the box. Default is `development`. 

- `port` - Port to use while running server. Default is 5000.

Eg.

```
$ python manage.py runserver --env production --port 5566
``` 

Runs the  server in `production` environment on port 5566.

You can  extend `manage.py` to support additional commands, parameters etc.

## Developing your application

*flask-quickstart* is built in such a way that you can easily integrate extensions, Blueprints, tests etc. The files are commented to help you out with this.



CHANGES
Changelog
=========

Changes between each release.

Version 1.1.2
-------------

Released on Jan 25, 2020

* Added .gtikeep file in app/templates/common, app/models folders.

Version 1.1.1
-------------

Released on October 28, 2019

* Added .gtikeep file in js, css and img folders in app/static.

Version 1.1.0
-------------

Released on October 17, 2019.

* Included tests folder and sample test file.
* Included 'testing' environment.
* TESTING is now set to False by default in 'development' environment.
* Included instance folder and example_config.py.
* The .gitignore file set to exclude all files in instance folder except example_config.py.


Version 1.0.0
-------------

Released on October 16th, 2019.

* First release of flask-quickstart
* Public API

tests/__init__.py
# This folder contains all your test files.
# These files are conventionally named test_*.

# It is not required that all your files need to be in this folder,
# you can create subfolders.
# For eg. all the tests for main Blueprint could go in the sub folder
# flask-quickstart/tests/main/

# There is a test_main.py included in this folder for reference.
# I'm using unnitest module from the standard library.
# There are other options available like pytest, nose etc.


tests/test_main.py
from app import create_app
import unittest

# How to run this file ?
# Go to the flask-quickstart directory
# ie. the directory above this one
# and execuute the following from the terminal
# python -m unittest tests.test_main

# ../flask-quickstart$ python -m unittest tests.test_main

# You IDE / editor will most likely have a feature/extension
# that automatically discovers and runs tests.


class TestMain(unittest.TestCase):
    def setUp(self):
        self.app = create_app('testing')
        self.client = self.app.test_client()

    def test_main(self):
        rv = self.client.get('/')
        assert rv.status == '200 OK'

    def test_404(self):
        rv = self.client.get('/unknownendpoint')
        self.assertEqual(rv.status, '404 NOT FOUND')


app/__init__.py
from flask import Flask
from config import config
# config module is available here
# https://github.com/miguelgrinberg/flasky/issues/154
# By default, the current directory for a process is the directory
# from where the application was started.

# Load all your extensions below ...
# Eg.
# from flask_sqlalchemy import SQLAlchemy
# from flask_bootstrap import Bootstrap
# db = SQLAlchemy()
# bootstrap = Bootstrap()


def create_app(config_name):
    app = Flask(__name__, instance_relative_config=True)
    app.config.from_object(config[config_name])
    config[config_name].init_app(app)

    # Configure the extensions below ...
    # Eg.
    # db.init_app(app)
    # bootstrap.init_app(app)

    # This is the main Blueprint
    from .main.views import main as main_blueprint
    app.register_blueprint(main_blueprint)
    # Register other Blueprints below ...

    return app


config.py
from flask import render_template


# Custom error pages
def page_not_found(e):
    return render_template('error/404.html'), 404


def internal_server_error(e):
    return render_template('error/500.html'), 500


class Production:
    DEBUG = False
    TESTING = False
    ENV = 'production'
    # This will be overwritten by the values in
    # instance/config.py
    SECRET_KEY = 'production_secret_key'

    @staticmethod
    def init_app(app):
        # flask-quickstart/instance/config.py
        app.config.from_pyfile('config.py', silent=True)
        app.register_error_handler(404, page_not_found)
        app.register_error_handler(500, internal_server_error)
        return


class Development:
    DEBUG = True
    TESTING = False
    ENV = 'development'
    SECRET_KEY = 'development_secret_key'

    @staticmethod
    def init_app(app):
        app.register_error_handler(404, page_not_found)
        return


class Testing:
    DEBUG = False
    TESTING = True
    ENV = 'testing'
    SECRET_KEY = 'testing_secret_key'

    @staticmethod
    def init_app(app):
        pass


config = {
    'production': Production,
    'development': Development,
    'testing': Testing
}
